Adding context menu and rendering options for certain files

dereklu888 opened the pull request:
Currently, only the basic context menu is implemented, I'll be adding rendering functionality soon.

dereklu888 commented:
Some issues I'll still need to clear up (not an exhaustive list, just some things I thought I'd note down):
- how to resolve if there are 2 links in the same row, one that goes to an internal page and another that goes to an external website (or just use the default browser context menu)
- downloading the libraries used and hosting it locally rather than pulling it from the internet
- whether there should be more options available in the context menu, like opening in a new window or others.

yarikoptic commented:
Thanks for posting preview to http://datasets-dev.datalad.org/ so I had chance to try it out.  Here is what I think we should do:

- Let's not overload right-click (so users can still resort to what browser provides them with) but prepend each path with ⋮ (in grey, may be there is a better symbol or icon to make it more obvious..) which would bring up the contextual menu
  - I think it is ok to keep "Copy Link" but I think "Open Link" isn't worth keeping there since that is the regular action happening on clicking the entry
- Remove "open in new tab" -- with above people could resort to the browser configuration/shortcuts
- For folders/subdatasets
 - there should be no "View in renderer" (for now at least)
 - if that folder is present locally (we have information since we do cross-out those which aren't), add "Open folder" and harmonize "Visit folder" to be also "Open folder".  If it could be made to just be an `<a>` within the menu - then they could again resort to browser features to open in new tab or not.  If it can't be `<a>` -- then ok, to not bother with "in new tab" version.
 - If there is an external URL available, add "Open external location"
- "View in Renderer" should be available only for files for which we have some renderer to use. ATM it is AFAIK only for .md files (but doesn't do anything yet)

yarikoptic commented on datalad_deprecated/resources/website/index.html:
```diff
@@ -29,6 +30,8 @@
 	<script type="text/javascript" charset="utf-8" src=".git/datalad/web/assets/js/jquery-2.2.4.js"></script>
 	<script type="text/javascript" charset="utf-8" src=".git/datalad/web/assets/js/jquery.dataTables-1.10.12.js"></script>
+	<script type="text/javascript" charset="utf-8" src="https://cdnjs.cloudflare.com/ajax/libs/jquery-contextmenu/2.9.2/jquery.contextMenu.min.js"></script>
+	<script type="text/javascript" charset="utf-8" src="https://cdnjs.cloudflare.com/ajax/libs/jquery-contextmenu/2.9.2/jquery.ui.position.min.js"></script>
```
we should carry and "deploy" copies of those along with the rest of .js.  Hopefully they do not need much of dependencies in turn

yarikoptic commented on datalad_deprecated/resources/website/index.html:
```diff
@@ -7,6 +7,7 @@
         <link rel="shortcut icon" href=".git/datalad/web/assets/images/favicon.ico?v=1">
         <link rel="stylesheet" type="text/css" href=".git/datalad/web/assets/css/jquery.dataTables-1.10.12.css">
         <link rel="stylesheet" type="text/css" href=".git/datalad/web/assets/css/main.css">
+        <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/jquery-contextmenu/2.9.2/jquery.contextMenu.min.css">
```
this one as well should be carried along

dereklu888 commented:
I implemented a 3dot menu for each row, with <a> tags for opening the folder and external link -- should be up on the datasets-dev. Quick question, still mulling about rendering, do you think the renderer should open the file in the current window (like replacing what is rendered where README.md usually is), or should it lead to an external link with just the rendered object (like viewing a file on github)? For the latter, I think it's doable without further infrastructure stuff/routing (which would make things less lean), but it might be less elegant.

Also, I implemented the context menu natively instead (let me know if there's any styling issues or bugs encountered), so I won't need to host new files for it.

yarikoptic commented:
re styling:
![image](https://user-images.githubusercontent.com/39889/127327515-d772fe4b-90da-4944-9b27-8a3f4fddf869.png)

Let's
- remove grey background (could be done here even though unrelated) behind that "[Open folder]", and make it in the color of the folder (of `func` to the left)
- context menu text: make consistent in font style (may be it is) and size (seems bigger) to the items in the table itself
- black text for context menu for some reason throws me off, not sure why ;)

Re renderers: let's first just add links to the remote resource like Bioimagesuite ones for `.nii(.gz)?$` files listed on https://github.com/dandi/dandiarchive/wiki/WiP:-dump-of-thoughts-on-%22services-registry%22 . Those, since regular links, could be opened by user in a new window or not as they like... although may be later we will provide a uniform new `target` so people could sequentially open different files in that viewer without breeding tabs - yet to decide.  ATM going https://bioimagesuiteweb.github.io/webapp/viewer.html?image=https://datasets-dev.datalad.org/dbic/QA/sub-emmet/ses-20180531/anat/sub-emmet_ses-20180531_acq-MPRAGE_T1w.nii.gz isn't working for me, but I will see if I just need to tune up CORS or report to bioimagesuite folks

yarikoptic commented:
woohoo -- I have fixed web server setup to provide certificate for datasets-dev so it could be served over https, and https://bioimagesuiteweb.github.io/webapp/viewer.html?image=https://datasets-dev.datalad.org/dbic/QA/sub-emmet/ses-20180531/anat/sub-emmet_ses-20180531_acq-MPRAGE_T1w.nii.gz works for me in brave browser (on chromium  site whines "Your browser does not support WEBGL (even v1). We can not proceeed.")

yarikoptic commented:
May be right away add that MetaCell/NWBExplorer link for .nwb files. I have made a sample file available under -dev in http://datasets-dev.datalad.org/?dir=/dandi/dandisets/000027/sub-RAT123 . Although ATM display there doesn't quite work (https://github.com/MetaCell/nwb-explorer/issues/261) but I hope it would get resolved soon.

dereklu888 commented:
I added render options for the .nii/.nii.gz and .nwb files, also tweaked the styling as per your changes (though the font for the context menu I just changed to a dark gray, if that's enough).

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -565,14 +600,48 @@ function directory(jQuery, md5) {
+      jQuery('table button').click(function(event) {
+        event.stopPropagation();
+        if (!jQuery(this).next().is(":visible")){
+          jQuery('.context-content:visible').hide();
```
looks like identical line here and in `else` -- move up outside of `if`. Code duplication is evil!

dereklu888 replied:
Oh wait, I do need both the lines, as the line after the hide() depends on the previous hide() statement (hide all others and show this one vs close everything including the currently open one).

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -565,14 +600,48 @@ function directory(jQuery, md5) {
+      //   console.log(jQuery(this).parent().parent().next().next().prop('href'));
+      //   copyToClipboard(jQuery(this).parent().parent().next().next().prop('href'));
+      //   jQuery('.context-content:visible').hide();
+      // });
```
not needed ? remove

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -530,6 +530,40 @@ function directory(jQuery, md5) {
+      }
+      else if (/.nwb$/.test(traverse.next)) {
+        menu +=  '<li class="context-option render"><a href="http://nwbexplorer.opensourcebrain.org/nwbfile=' + link.href + '">View in Renderer</a></li>';
+      }
```
Let's not postpone properly separating the "spec"  of "renders" from code generating the menu. (and again -- code duplication is evil! ;))

- take that https://github.com/dandi/dandiarchive/wiki/WiP:-dump-of-thoughts-on-%22services-registry%22 and make it into a data structure on top of the script, e.g. "EXTERNAL_SERVICES" (just exclude/comment out "NWB Widgets (Binder):" for now).  Could be simplified or just harmonized (always a list) if desired to avoid that single item vs list, up to you.  Can do simple string replacement of `{url}` with actual url for now, since we have no other values to pass and you didn't want to use string interpolations 
- have a function given an item (`traverse.next` here?) return a dict of "name: URL" which you render here in the loop as `<li` after "External Services" section of a kind (just e.g. a bolded text in the menu, doesn't need to do anything upon click). Note -- name should be `Bioimagesuite/Viewer` etc, not just "View in Renderer" since a single type of file could have various.

Then later we might even have such data structure defined/fetched externally and code here would be ready to handle any new ones possibly without even redeploying anything.

yarikoptic replied:
also -- make `traverse.next` called ones and assigned near `var traverse` since I  guess (didn't check) it would run server queries then for each invocation, which could be avoided.

dereklu888 replied:
I added a data structure that should do that (though I still need to test more and check about some of the other file types, like directory) -- should be up on datasets-dev. I'll add the function you mentioned in point 2 soon.

As for the `traverse.next`, I think accessing that isn't a server query, only the initial `var traverse = clickhandler(data)` is.

yarikoptic reviewed (changes requested):
Looks sweet, thank you !  But now let's make it "flexible" by separating spec of "renders" (which might not even be renderers but some processing etc, so let's call "External Services")  from code. Left more specific comments

yarikoptic commented:
also styling wise, a minor comment
![image](https://user-images.githubusercontent.com/39889/127657637-3cfc728b-793d-48c9-b963-f785fdc43155.png)

if three dots could be aligned to text better (didn't check if font style/size differs from the text?) would be nice.  Now it is bigger and a bit higher than text.

dereklu888 commented:
As for the styling -- sorry, I thought you meant the font styling of the menu options themselves, not the 3 dots. I'll try to finish that up next.

dereklu888 commented:
Alright, I updated the formatting of the 3dot menu, added an External Section separator in the menu, and added the rendering function -- let me know if the format of the rendering data looks good, I'm not sure yet if I should keep the extensions and use them as a key in a map, let me know your thoughts.

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -594,4 +732,4 @@ function clearCache() {
 
 window.addEventListener('beforeunload', clearCache, false);
-*/
+*/
```
please don't delete trailing new line here -- unrelated to the changes in PR and in general "good to have"

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -585,6 +691,38 @@ function directory(jQuery, md5) {
+
+      validRenderers.push({
+        "name": renderer.name,
+        "url": renderer.endpoint + link.href,
```
eventually we should make it a proper "interpolation" or alike, but for now as long as data structure defining renderers is "built-in", ok - let it be just an append until it can no longer ;)

yarikoptic reviewed (changes requested):
Looks very nice, thank you!  Left just one stylistic comment.  Another one:

is there an easy way to have '...' to not disappear whenever mouse leaves that row/menu if menu is still shown? if not -- that is ok.

And then I think we should proceed/merge/deploy and deal with the rest later on.  Unless may be you are planing to add a basic test or two ;)

dereklu888 commented:
For styling, I opted to make the External Services separator the same left indent as the copy/other operations, and increased the indent of the renderers (I just set it as a hard 15px). I added a few tests in (though they are currently fitted to the 3 rendering options available, might need to go update this when new rendering options are added?).

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -565,14 +631,41 @@ function directory(jQuery, md5) {
+      
       jQuery('#directory_filter').prepend('<a id="folder-link" href="'+curdir+'">'+
-                                          '<span class="visit-folder">[Visit folder]</span>'+
+                                          '<span class="visit-folder ' + getNodeType(jQuery, md5, rawCrumbs.join('/')) + '">[Open Folder]</span>'+
```
note: `visit-folder` custom css was removed, so class has no effect here, but I am ok to keep it around

yarikoptic commented on datalad_deprecated/resources/website/assets/js/main.js:
```diff
@@ -585,6 +678,38 @@ function directory(jQuery, md5) {
+function getRenderers(filename) {
+  validRenderers = []
+
+  EXTERNAL_RENDERERS.forEach(function(renderer) {
```
I was on the edge on this, but for the clarity lets rename `EXTERNAL_RENDERERS` to `EXTERNAL_SERVICES` since that is what we call them in the menu -- consistency is good! and thus `getRenderers` into `getExternalServices` and anything else which mentions "render" whenever it might be just giving away the file for torture of a kind we have no clue about

yarikoptic reviewed (changes requested):
very minor -- just do renames for consistency with menu and reality

yarikoptic commented on datalad_deprecated/resources/website/tests/test.html:
```diff
@@ -261,6 +262,24 @@
       });
 
+      
+      QUnit.test("Context Menu Renderers", function(assert) {
```
I guess forgotten here?

dereklu888 replied:
Ach, sorry about that, should be fixed now.

yarikoptic commented:
Thank you @dereklu888 !  Let's proceed with this one.  I will also deploy current state of the web ui to the main instance
